File: tools/gen_unicode.py
#!/usr/bin/env python3
"""Build the checked-in Unicode 15.0.0 runtime tables from Python's pinned UCD."""

from __future__ import annotations

import argparse
import hashlib
import struct
import sys
import unicodedata
from pathlib import Path


VERSION = "15.0.0"
MAGIC = b"PEONYU15"
VERSION_BYTES = VERSION.encode("ascii")
RANGE = struct.Struct("<IIH")
MAPPING_HEAD = struct.Struct("<I")
MAPPING_MAP = struct.Struct("<B3I")
SIMPLE = struct.Struct("<I")
RANGE_FLAGS = {
    "alphabetic": 1 << 0,
    "alnum": 1 << 1,
    "decimal": 1 << 2,
    "digit": 1 << 3,
    "numeric": 1 << 4,
    "whitespace": 1 << 5,
    "lower": 1 << 6,
    "upper": 1 << 7,
    "title": 1 << 8,
    "cased": 1 << 9,
    "case_ignorable": 1 << 10,
    "title_ignorable": 1 << 11,
}
CASE_IGNORABLE_WORD_BREAK = {
    0x0027, 0x00AD, 0x00B7, 0x0387, 0x05F4, 0x2019, 0x2027,
    0xFE13, 0xFE55, 0xFF07, 0xFF1A, 0xFF65,
}
IGNORABLE_CATEGORIES = {"Mn", "Me", "Cf", "Lm", "Sk"}


def codepoints(text: str) -> tuple[int, ...]:
    return tuple(map(ord, text))


def full_mapping(character: str, operation: str) -> tuple[int, ...]:
    mapped = codepoints(getattr(character, operation)())
    return () if mapped == (ord(character),) else mapped


def simple_mapping(cp: int, mapping: tuple[int, ...], operation: str) -> int:
    if len(mapping) == 1:
        return mapping[0]
    # UnicodeData's simple lowercase for dotted capital I is i; the full UCD
    # mapping expands to i + COMBINING DOT ABOVE.
    if cp == 0x0130 and operation == "lower":
        return 0x0069
    return cp


def simple_fold(cp: int, casefold: tuple[int, ...], lower: tuple[int, ...]) -> int:
    if len(casefold) == 1:
        return casefold[0]
    # For capital sharp S the full fold is "ss", while the simple fold is ß.
    if len(lower) == 1:
        return lower[0]
    return cp


def char_flags(cp: int, char: str) -> int:
    category = unicodedata.category(char)
    flags = 0
    flags |= RANGE_FLAGS["alphabetic"] if char.isalpha() else 0
    flags |= RANGE_FLAGS["alnum"] if char.isalnum() else 0
    flags |= RANGE_FLAGS["decimal"] if char.isdecimal() else 0
    flags |= RANGE_FLAGS["digit"] if char.isdigit() else 0
    flags |= RANGE_FLAGS["numeric"] if char.isnumeric() else 0
    flags |= RANGE_FLAGS["whitespace"] if char.isspace() else 0
    flags |= RANGE_FLAGS["lower"] if char.islower() else 0
    flags |= RANGE_FLAGS["upper"] if char.isupper() else 0
    flags |= RANGE_FLAGS["title"] if char.istitle() else 0
    if flags & (RANGE_FLAGS["lower"] | RANGE_FLAGS["upper"] | RANGE_FLAGS["title"]):
        flags |= RANGE_FLAGS["cased"]
    if category in IGNORABLE_CATEGORIES or cp in CASE_IGNORABLE_WORD_BREAK:
        flags |= RANGE_FLAGS["case_ignorable"]
    if category in IGNORABLE_CATEGORIES:
        flags |= RANGE_FLAGS["title_ignorable"]
    return flags


def build_blob() -> tuple[bytes, str]:
    if unicodedata.unidata_version != VERSION:
        raise RuntimeError(
            f"expected Unicode {VERSION}; Python provides {unicodedata.unidata_version}"
        )

    ranges: list[tuple[int, int, int]] = []
    mappings: list[bytes] = []
    input_digest = hashlib.sha256()
    range_start: int | None = None
    previous = -1
    previous_flags = 0

    for cp in range(0x110000):
        char = chr(cp)
        flags = char_flags(cp, char)
        full = (
            full_mapping(char, "lower"),
            full_mapping(char, "upper"),
            full_mapping(char, "title"),
            full_mapping(char, "casefold"),
        )
        simple = (
            simple_mapping(cp, full[0], "lower"),
            simple_mapping(cp, full[1], "upper"),
            simple_mapping(cp, full[2], "title"),
            simple_fold(cp, full[3], full[0]),
        )

        # Fingerprint the complete relevant UCD projection, including unchanged
        # scalars. This is stable across machines that expose the same UCD.
        row = bytearray(6 + 4 * (MAPPING_MAP.size + SIMPLE.size))
        struct.pack_into("<IH", row, 0, cp, flags)
        offset = 6
        for full_map, simple_map in zip(full, simple, strict=True):
            padded = (*full_map[:3], 0, 0, 0)[:3]
            MAPPING_MAP.pack_into(row, offset, len(full_map), *padded)
            offset += MAPPING_MAP.size
            SIMPLE.pack_into(row, offset, simple_map)
            offset += SIMPLE.size
        input_digest.update(row)

        if flags and range_start is not None and flags == previous_flags and cp == previous + 1:
            previous = cp
        else:
            if range_start is not None:
                ranges.append((range_start, previous, previous_flags))
            range_start = cp if flags else None
            previous = cp
            previous_flags = flags

        if any(full_map for full_map in full) or any(simple_map != cp for simple_map in simple):
            record = bytearray(MAPPING_HEAD.size + 4 * (MAPPING_MAP.size + SIMPLE.size))
            MAPPING_HEAD.pack_into(record, 0, cp)
            offset = MAPPING_HEAD.size
            for full_map, simple_map in zip(full, simple, strict=True):
                padded = (*full_map[:3], 0, 0, 0)[:3]
                MAPPING_MAP.pack_into(record, offset, len(full_map), *padded)
                offset += MAPPING_MAP.size
                SIMPLE.pack_into(record, offset, simple_map)
                offset += SIMPLE.size
            mappings.append(bytes(record))

    if range_start is not None:
        ranges.append((range_start, previous, previous_flags))

    input_hash = input_digest.digest()
    header = MAGIC + VERSION_BYTES + struct.pack("<II", len(ranges), len(mappings)) + input_hash
    blob = bytearray(header)
    for first, last, flags in ranges:
        blob.extend(RANGE.pack(first, last, flags))
    for record in mappings:
        blob.extend(record)
    return bytes(blob), input_digest.hexdigest()


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--output",
        type=Path,
        default=Path(__file__).resolve().parents[1] / "data" / "unicode-15.0.bin",
    )
    parser.add_argument("--check", action="store_true", help="verify the checked-in blob without writing")
    args = parser.parse_args()

    try:
        blob, input_hash = build_blob()
    except RuntimeError as error:
        print(error, file=sys.stderr)
        return 2

    output_hash = hashlib.sha256(blob).hexdigest()
    details = (
        f"Unicode {VERSION}; projection-sha256={input_hash}; "
        f"blob-sha256={output_hash}; bytes={len(blob)}"
    )
    if args.check:
        try:
            existing = args.output.read_bytes()
        except OSError as error:
            print(f"cannot read {args.output}: {error}", file=sys.stderr)
            return 2
        if existing != blob:
            print(f"Unicode blob does not match generator output: {args.output}", file=sys.stderr)
            print(details, file=sys.stderr)
            return 1
        print(f"match: {args.output} ({details})")
        return 0

    args.output.parent.mkdir(parents=True, exist_ok=True)
    args.output.write_bytes(blob)
    print(f"wrote: {args.output} ({details})")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: tools/gen_unicode.mjs
#!/usr/bin/env node
// Build Peony's Unicode 15 table directly from pinned official UCD files.
import { createHash } from 'node:crypto';
import { readFile, writeFile, mkdir } from 'node:fs/promises';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const root = fileURLToPath(new URL('../', import.meta.url));
const ucd = path.join(root, 'data', 'ucd-15.0.0');
const expectedFiles = {
  'CaseFolding.txt': 'cdd49e55eae3bbf1f0a3f6580c974a0263cb86a6a08daa10fbf705b4808a56f7',
  'DerivedCoreProperties.txt': 'd367290bc0867e6b484c68370530bdd1a08b6b32404601b8c7accaf83e05628d',
  'SpecialCasing.txt': '78b29c64b5840d25c11a9f31b665ee551b8a499eca6c70d770fcad7dd710f494',
  'UnicodeData.txt': '806e9aed65037197f1ec85e12be6e8cd870fc5608b4de0fffd990f689f376a73',
  'Unihan_NumericValues.txt': '42289ff99564cf17c3c95938744c2f690452b704a6d076d5372c4571c3cb14f6',
};
const expectedProjection = '4823dcf0e05d64d5c2fa3e4de8f23935959b9c86fd4d89f41a7461280ffed10b';
const expectedBlob = '10a4fd50df393d992424d2102e48e39c58dc6449a7a9cae0dcc5708683848019';
const args = process.argv.slice(2);
let check = false;
let output = path.join(root, 'data', 'unicode-15.0.bin');
for (let index = 0; index < args.length; index++) {
  if (args[index] === '--check') check = true;
  else if (args[index] === '--output' && args[index + 1]) output = path.resolve(args[++index]);
  else throw new Error('usage: node tools/gen_unicode.mjs [--check] [--output FILE]');
}
const sha = (bytes) => createHash('sha256').update(bytes).digest('hex');
const sources = {};
for (const [name, expected] of Object.entries(expectedFiles)) {
  const bytes = await readFile(path.join(ucd, name));
  if (sha(bytes) !== expected) throw new Error(`pinned official UCD hash mismatch: ${name}`);
  sources[name] = bytes.toString('utf8');
}
const SIZE = 0x110000;
const category = Array(SIZE).fill('Cn');
const bidi = Array(SIZE).fill('');
const decimal = new Uint8Array(SIZE);
const digit = new Uint8Array(SIZE);
const numeric = new Uint8Array(SIZE);
const alpha = new Uint8Array(SIZE);
const lowerProperty = new Uint8Array(SIZE);
const upperProperty = new Uint8Array(SIZE);
const simple = new Map();
function scalar(text) { return Number.parseInt(text, 16); }
function codepoints(text) { return text.trim() ? text.trim().split(/\s+/).map(scalar) : []; }
function range(text) {
  const endpoints = text.trim().split('..');
  const first = scalar(endpoints[0]);
  return [first, endpoints[1] ? scalar(endpoints[1]) : first];
}
let firstRange = null;
for (const raw of sources['UnicodeData.txt'].trimEnd().split(/\r?\n/)) {
  const fields = raw.split(';');
  const cp = scalar(fields[0]);
  const name = fields[1];
  if (name.endsWith(', First>')) { firstRange = { cp, fields }; continue; }
  const start = name.endsWith(', Last>') ? firstRange.cp : cp;
  if (name.endsWith(', Last>') && (!firstRange || firstRange.fields[2] !== fields[2])) throw new Error('bad UCD First/Last range');
  for (let target = start; target <= cp; target++) {
    category[target] = fields[2];
    bidi[target] = fields[4];
    decimal[target] = fields[6] ? 1 : 0;
    digit[target] = fields[7] ? 1 : 0;
    numeric[target] = fields[8] ? 1 : 0;
  }
  if (name.endsWith(', Last>')) { firstRange = null; continue; }
  simple.set(cp, {
    upper: fields[12] ? [scalar(fields[12])] : [],
    lower: fields[13] ? [scalar(fields[13])] : [],
    title: fields[14] ? [scalar(fields[14])] : [],
  });
}
if (firstRange) throw new Error('unterminated UCD First/Last range');
for (const raw of sources['Unihan_NumericValues.txt'].split(/\r?\n/)) {
  const content = raw.split('#', 1)[0].trim();
  if (!content) continue;
  const fields = content.split('\t');
  if (fields.length !== 3 || !fields[0].startsWith('U+')) throw new Error('invalid Unihan numeric row');
  numeric[scalar(fields[0].slice(2))] = 1;
}
for (const raw of sources['DerivedCoreProperties.txt'].split(/\r?\n/)) {
  const content = raw.split('#', 1)[0].trim();
  if (!content) continue;
  const [rangeText, propertyText] = content.split(';');
  const property = propertyText.trim();
  const target = property === 'Alphabetic' ? alpha : property === 'Lowercase' ? lowerProperty : property === 'Uppercase' ? upperProperty : null;
  if (!target) continue;
  const [first, last] = range(rangeText);
  target.fill(1, first, last + 1);
}
const special = new Map();
for (const raw of sources['SpecialCasing.txt'].split(/\r?\n/)) {
  const content = raw.split('#', 1)[0].trim();
  if (!content) continue;
  const fields = content.split(';').map((field) => field.trim());
  if (fields[4]) continue; // locale/context rules do not affect one-scalar projection
  special.set(scalar(fields[0]), {
    lower: codepoints(fields[1]),
    title: codepoints(fields[2]),
    upper: codepoints(fields[3]),
  });
}
const folds = new Map();
for (const raw of sources['CaseFolding.txt'].split(/\r?\n/)) {
  const content = raw.split('#', 1)[0].trim();
  if (!content) continue;
  const fields = content.split(';').map((field) => field.trim());
  if (fields[1] !== 'C' && fields[1] !== 'F') continue;
  const cp = scalar(fields[0]);
  const previous = folds.get(cp);
  if (!previous || fields[1] === 'F') folds.set(cp, { status: fields[1], mapping: codepoints(fields[2]) });
}
const ignoredCategories = new Set(['Mn', 'Me', 'Cf', 'Lm', 'Sk']);
const breakIgnorable = new Set([0x27, 0xad, 0xb7, 0x387, 0x5f4, 0x2019, 0x2027, 0xfe13, 0xfe55, 0xff07, 0xff1a, 0xff65]);
const whitespaceBidi = new Set(['WS', 'B', 'S']);
const ranges = [];
const mappings = [];
const digest = createHash('sha256');
const fingerprintRow = Buffer.alloc(74);
let rangeStart = -1;
let previous = -1;
let previousFlags = 0;
function pushRange() {
  if (rangeStart < 0) return;
  const record = Buffer.alloc(10);
  record.writeUInt32LE(rangeStart, 0);
  record.writeUInt32LE(previous, 4);
  record.writeUInt16LE(previousFlags, 8);
  ranges.push(record);
}
for (let cp = 0; cp < SIZE; cp++) {
  const cat = category[cp];
  const lower = !!lowerProperty[cp];
  const upper = !!upperProperty[cp];
  const title = upper || cat === 'Lt';
  const alphabetic = cat.startsWith('L');
  const isNumeric = !!numeric[cp];
  const ignored = ignoredCategories.has(cat);
  let flags = 0;
  if (alphabetic) flags |= 1;
  if (alphabetic || isNumeric) flags |= 2;
  if (decimal[cp]) flags |= 4;
  if (digit[cp]) flags |= 8;
  if (isNumeric) flags |= 16;
  if (cat === 'Zs' || whitespaceBidi.has(bidi[cp])) flags |= 32;
  if (lower) flags |= 64;
  if (upper) flags |= 128;
  if (title) flags |= 256;
  if (lower || upper || title) flags |= 512;
  if (ignored || breakIgnorable.has(cp)) flags |= 1024;
  if (ignored) flags |= 2048;

  const simpleCase = simple.get(cp);
  const specialCase = special.get(cp);
  const fold = folds.get(cp);
  const full = [
    specialCase?.lower ?? simpleCase?.lower ?? [],
    specialCase?.upper ?? simpleCase?.upper ?? [],
    specialCase?.title ?? simpleCase?.title ?? [],
    fold?.mapping ?? [],
  ].map((mapping) => mapping.length === 1 && mapping[0] === cp ? [] : mapping);
  const single = [
    full[0].length === 1 ? full[0][0] : cp === 0x130 ? 0x69 : cp,
    full[1].length === 1 ? full[1][0] : cp,
    full[2].length === 1 ? full[2][0] : cp,
    full[3].length === 1 ? full[3][0] : full[0].length === 1 ? full[0][0] : cp,
  ];
  fingerprintRow.writeUInt32LE(cp, 0);
  fingerprintRow.writeUInt16LE(flags, 4);
  for (let operation = 0; operation < 4; operation++) {
    const offset = 6 + operation * 17;
    fingerprintRow.writeUInt8(full[operation].length, offset);
    for (let scalarIndex = 0; scalarIndex < 3; scalarIndex++) fingerprintRow.writeUInt32LE(full[operation][scalarIndex] ?? 0, offset + 1 + scalarIndex * 4);
    fingerprintRow.writeUInt32LE(single[operation], offset + 13);
  }
  digest.update(fingerprintRow);
  if (flags && rangeStart >= 0 && flags === previousFlags && cp === previous + 1) previous = cp;
  else {
    pushRange();
    rangeStart = flags ? cp : -1;
    previous = cp;
    previousFlags = flags;
  }
  if (full.some((mapping) => mapping.length) || single.some((mapping) => mapping !== cp)) {
    const record = Buffer.alloc(72);
    record.writeUInt32LE(cp, 0);
    fingerprintRow.copy(record, 4, 6, 74);
    mappings.push(record);
  }
}
pushRange();
const fingerprint = digest.digest('hex');
const header = Buffer.alloc(54);
header.write('PEONYU15', 0, 'ascii');
header.write('15.0.0', 8, 'ascii');
header.writeUInt32LE(ranges.length, 14);
header.writeUInt32LE(mappings.length, 18);
Buffer.from(fingerprint, 'hex').copy(header, 22);
const blob = Buffer.concat([header, ...ranges, ...mappings]);
const blobHash = sha(blob);
if (fingerprint !== expectedProjection || blobHash !== expectedBlob) {
  throw new Error(`Unicode 15 projection differs: projection=${fingerprint}, blob=${blobHash}, ranges=${ranges.length}, mappings=${mappings.length}`);
}
if (check) {
  const existing = await readFile(output);
  if (!existing.equals(blob)) throw new Error(`Unicode blob does not match official UCD projection: ${output}`);
  process.stdout.write(`match: ${output} (Unicode 15.0.0; projection-sha256=${fingerprint}; blob-sha256=${blobHash}; bytes=${blob.length})\n`);
} else {
  await mkdir(path.dirname(output), { recursive: true });
  await writeFile(output, blob);
  process.stdout.write(`wrote: ${output} (Unicode 15.0.0; projection-sha256=${fingerprint}; blob-sha256=${blobHash}; bytes=${blob.length})\n`);
}
